src/main/java/net/dries007/tfc/objects/container/ContainerKnapping.java

```diff
@@ -79,14 +81,39 @@ public void onButtonPress(int buttonID, @Nullable NBTTagCompound extraNBT)
         }
     }
 
+    @Nonnull
+    @Override
+    public ItemStack slotClick(int slotID, int dragType, ClickType clickType, EntityPlayer player)
+    {
+        Slot slot = inventorySlots.get(0);
+        if (slot.slotNumber == slotID && slot.getHasStack() && type.consumeAfterComplete() && !player.isCreative())
+        {
+            ItemStack stack = Helpers.consumeItem(this.stack, type.getAmountToConsume());
+            if (isOffhand)
+            {
+                player.setHeldItem(EnumHand.OFF_HAND, stack);
+            }
+            else
+            {
+                player.setHeldItem(EnumHand.MAIN_HAND, stack);
+            }
+        }
+        return super.slotClick(slotID, dragType, clickType, player);
+    }
+
     @Override
     public void onContainerClosed(EntityPlayer player)
     {
         Slot slot = inventorySlots.get(0);
         ItemStack stack = slot.getStack();
         if (!stack.isEmpty())
         {
-            player.addItemStackToInventory(stack);
+            if (!player.world.isRemote)
+            {
+                // Player#addItemStackToInventory(stack) returned true even when inventory was full
+                // Since the item is immediately picked up when possible, this works like we want
+                Helpers.spawnItemStack(player.world, player.getPosition(), stack);
```

Comment:
```suggestion
                ItemHandlerHelper.giveItemToPlayer(player, stack);
```
And the comments can be removed, as they are unnecessary,